lib/gooddata/lcm/actions/synchronize_user_filters.rb

```diff
@@ -159,46 +166,50 @@ def call(params)
               domain_clients.select! { |c| segments_filter.include?(c.segment_uri) }
             end
 
-            working_client_ids = []
+            working_client_ids = ThreadSafe::Array.new
 
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            filters.group_by { |u| u[multiple_projects_column] }.flat_pmap do |client_id, new_filters|
               fail "Client id cannot be empty" if client_id.blank?
-              c = domain.clients(client_id)
-              if params.segments_filter && !segments_filter.include?(c.segment_uri)
-                puts "Client #{client_id} is outside segments_filter #{params.segments_filter}"
+              segment_client = domain_clients.find { |c| c.client_id.to_s == client_id.to_s }
+              if params.segments_filter && !segments_filter.include?(segment_client.segment_uri)
+                semaphore.synchronize { params.gdc_logger.warn("Client #{client_id} is outside segments_filter #{params.segments_filter}") }
                 next
               end
-              project = c.project
-              fail "Client #{client_id} does not have project." unless project
-              working_client_ids << client_id
-              filters_to_load = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
-              puts "Synchronizing #{filters_to_load.count} filters in project #{project.pid} of client #{client_id}"
-              project.add_data_permissions(filters_to_load, run_params)
+              client_project = segment_client.project
+              fail "Client #{client_id} does not have project." unless client_project
+              working_client_ids << client_id.to_s
+              Thread.current['filters_to_load'] = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
+              semaphore.synchronize do
+                params.gdc_logger.info("Synchronizing #{Thread.current['filters_to_load'].count} filters in project #{client_project.pid} of client #{client_id}")
+              end
+              client_project.add_data_permissions(Thread.current['filters_to_load'], run_params)
             end
 
             unless run_params[:do_not_touch_filters_that_are_not_mentioned]
-              domain_clients.each do |c|
-                next if working_client_ids.include?(c.client_id)
+              domain_clients.peach do |c|
+                next if working_client_ids.include?(c.client_id.to_s)
```

Comment:
Similar code found in synchronize_user_filters.rb:190, synchronize_users.rb:276 (mass = 44)